File: Other Data Structures_qstackheap/le208_implementTrie.py
"""

Implement a trie with insert, search, and startsWith methods.
Note:
You may assume that all inputs are consist of lowercase letters a-z.

"""
class TrieNode:
    def __init__(self):
        self.childs = {}
        self.isWord = False

class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        node = self.root
        for letter in word:
           child = node.childs.get(letter)
           if child is None:
               child = TrieNode()
               node.childs[letter] = child
           node = child
        child.isWord = True
        

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        node = self.root
        for letter in word:
            if not node.childs.get(letter):
                return False
            node = node.childs[letter]
        return node.isWord
            

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        node = self.root
        for letter in prefix:
            if not node.childs.get(letter):
                return False
            node = node.childs[letter]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix) # start with means start from the root to the end of prefix

File: JavaPartI/le208_implementTrie.java
// hashmap version
class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>(); 
    boolean isWord;
    TrieNode() {

    }   
}
public class Trie {
    private TrieNode root; 
    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode cur = root;
        for (char ch : word.toCharArray()) {
            if (cur.children.containsKey(ch)) {
                cur = cur.children.get(ch);
            } else {
                TrieNode newnode = new TrieNode();
                cur.children.put(ch, newnode);
                cur = newnode;
            }
        }
        cur.isWord = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode cur = root;
        for (char ch : word.toCharArray()) {
            if (cur.children.containsKey(ch)) {
                cur = cur.children.get(ch);
            } else {
                return false;
            }
        }
        return cur.isWord;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode cur = root;
        for (char ch : prefix.toCharArray()) {
            if (cur.children.containsKey(ch)) {
                cur = cur.children.get(ch);
            } else {
                return false;
            }
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
